packages/compiler-sfc/src/script/defineModel.ts

```diff
@@ -120,12 +120,16 @@ export function genModelProps(ctx: ScriptCompileContext) {
     if (runtimeTypes) {
       const hasUnknownType = runtimeTypes.includes(UNKNOWN_TYPE)
 
-      runtimeTypes = runtimeTypes.filter(el => {
-        if (el === UNKNOWN_TYPE) return false
-        return isProd
-          ? el === 'Boolean' || (el === 'Function' && options)
-          : true
-      })
+      runtimeTypes = runtimeTypes.filter(el => el !== UNKNOWN_TYPE)
+
+      if (
+        isProd &&
+        !runtimeTypes.some(
+          el => el === 'Boolean' || (el === 'Function' && options)
+        )
+      ) {
+        runtimeTypes = []
+      }
```

Comment:
This code doesn't seem to solve #9587 very well. Do you only need to keep Boolean, String, and Function when there are multiple types

Reply:
I tried my code in your PR. The unit test is passed and nothing changed, comparing your implementation.

Reply:
Sry, I didn't write the test case accurately, I rewrote it according to the issue #9587 .

Reply:
The `Number` type is not necessary for production mode. It will be ignored since the type check is skipped in production mode. Only boolean, function (if default value exists), and string (if boolean type exists) are necessary for production mode and will affect runtime behavior.